Improve annotation import UI

hudson-newey opened the pull request:
# Improve annotation import UI

## Changes

- Fixes directory structure of annotation import pages. They were previously located directly under the `components/import_annotations/` directory when our convention is to place them under the `components/import_annotations/pages` directory
- Add warning to the menu widgets if files have not been committed
- Add menu widgets explaining how to fix errors
	- Add card explaining how to fix event errors which contains a hint that more descriptive errors can be found in the table.
	- Card explaining duplicate file errors.
	- Unsupported format error card
	- Internal server error card
- Change branding of "Batch Import Annotations" to "Import Annotations"
	- Updates page title
	- Updates breadcrumbs
	- Updates menu item text
	- Updates URL slug from `batch_annotations` to `import_annotations`
- Add gradient if ngx-datatable's are horizontally scrollable
- After creating an annotation import, users are now navigated to the file upload page instead of the "details" page that shows the files and events associated with the import.
- Bumps the `@ecoacoustics/web-component` version
	- Adds support for MacOS
	- Fixes indicator line going offscreen
	- Fixes indicator line moving if audio does not play
	- Some performance improvements
- Adds a "compact" configuration for the ngx-datatable
- Moves the annotation import menu link to the project details page
- `baw-error-card` component now fills its `:host` container. Meaning that changing the size of the `baw-error-card` selector now changes the size of the errors contained.

### Bug fixes

- Fixes the "Download Annotations" menu item being visible even if the user is not logged in. (previously visible on the project, region, site, and point detail pages)

## Issues

Fixes: #2297

## Visual Changes

![image](https://github.com/user-attachments/assets/c74e4c30-f161-4228-ba06-6c528cd6a772)

_New "Instructions" widget that is displayed before any action is taken_

---

![image](https://github.com/user-attachments/assets/5566023f-2035-4542-b497-d8ff0cad2d46)

_Information card explaining that a valid file has not been committed to the database until the "Import Annotations" button is clicked._

---

![image](https://github.com/user-attachments/assets/bee69bbb-b00f-4405-b4fc-36e1223cf1af)

_Information cards explaining how to fix errors in the uploaded datasets. Note that errors now fill up the entire cell._

## Final Checklist

- [x] Assign reviewers if you have permission
- [x] Assign labels if you have permission
- [x] Link issues related to PR
- [x] Ensure project linter is not producing any warnings (`npm run lint`)
- [x] Ensure build is passing on all browsers (`npm run test:all`)
- [ ] Ensure CI build is passing
- [ ] Ensure docker container is passing ([docs](https://github.com/QutEcoacoustics/workbench-client#docker))

hudson-newey committed:
Navigate to add annotations on import creation

hudson-newey committed:
Simplify naming for annotation import title

Change page title and menu label from "batch import anntoations" to
"import annotations".

hudson-newey committed:
Fix view button going to add annotations page

hudson-newey committed:
Add fase-overflow and information cards

hudson-newey committed:
Non-reactive demo

hudson-newey committed:
Fix annotation download menu predicates

hudson-newey committed:
Bump web components version

- Adds support for MacOS, Firefox, and Safari
- Fixes desync of currentTime and playback time (many fixes to indicator
  line)
- Layout performance significantly improved

hudson-newey committed:
Fix directory structure for annotation import

It was not the same as the rest of the client. I have now fixed the
directory structure so that it matches the rest of the client.

hudson-newey committed:
Basic shared state with import-annotation.service

[skip ci]

hudson-newey committed:
Initial version with working instructions

[skip ci]

hudson-newey committed:
Fix view model type leakage

hudson-newey committed:
Use route provider for annotation service

hudson-newey committed:
Start tests for instruction component

hudson-newey committed:
Improve ownership of signal instances

Consumers of the import annotation service can no longer directly modify
the signal.
Instead writers request a signal lease that they can update and
consumers connect to a non-writable signal.

hudson-newey committed:
Add support for unsupported content types

hudson-newey committed:
Remove missing tag warning

No one actually asked for this. I thought that it'd be simple and useful
to add, but it turned out that it's quite hard to make a good solution
to extract warnings.

This is because warnings are generated on a per-event basis, meaning
that we'd have to process the entire uploaded file client side (not
very performant).

hudson-newey committed:
Finish service tests

hudson-newey committed:
Add contact us links to instruction cards

hudson-newey committed:
Disable hmr in dev

HMR was not working nicely with dynamic imports and
getClientBoundingRect.

I have temporarily disabled it until hmr support is more stable.

hudson-newey committed:
Improve copy for instruction widgets

hudson-newey committed:
Add jsdoc comments for import-annotation.service

hudson-newey committed:
Fix instruction tests

hudson-newey committed:
Self review

hudson-newey committed:
Rename instructions component

hudson-newey committed:
Use hr elements for error card seperation

It has better semantics (semantic html), better accessability, and means
that consistent styling is easier.

hudson-newey committed:
Fix hard coded x-fade-overflow

hudson-newey committed:
Add parameters for x-fade-overflow

hudson-newey committed:
Add pointer events

hudson-newey committed:
Change ALL datatables to compact tables

hudson-newey committed:
Add compact datatable config

hudson-newey committed:
Use compact datatable for annotation import

hudson-newey committed:
Revert formatting changes

atruskie commented on src/app/components/import-annotations/widgets/annotation-import-issue.component.html:
```diff
@@ -0,0 +1,104 @@
+
+<p>
+  You can use this interface to upload annotations from human or machine
+  generated datasets.
```
```suggestion
  This page allows you to import annotations from human or machine
  generated datasets.
```

atruskie commented on src/app/components/import-annotations/widgets/annotation-import-issue.component.html:
```diff
@@ -0,0 +1,104 @@
+    </h6>
+
+    <strong>Your files have not been saved</strong>. Please review your uploaded
+    files, and click "Import Annotations" to save your annotations.
```
```suggestion
    <strong>Your files have not been imported yet</strong>. 
    Check the preview table and make sure we've
    correctly parsed your data.
```

atruskie commented on src/app/components/import-annotations/widgets/annotation-import-issue.component.html:
```diff
@@ -0,0 +1,104 @@
+</p>
+
+@if (hasUncommittedFiles()) {
+  <div class="uncommitted-file-warning alert alert-warning">
```
i think this should be an info rather than a warning

hudson-newey replied:
![image](https://github.com/user-attachments/assets/8f36d5fc-4796-4e6f-84df-3e6b8a8622a2)

Opinion?

Should it still say warning? Should it still have an exclamation mark?

---

Here's an alternative that I mocked up that has the "Information" heading, and a `circle-exclamation` icon.

![image](https://github.com/user-attachments/assets/b15ba802-afa9-425f-8ce7-cedd93ffcb53)

atruskie replied:
Don't we have stock components for these alerts? 

Maybe we don't need the _Information_/_Warning_ title at all. The first sentence of the paragraph can just be the title.

hudson-newey replied:
We don't have stock components for these alerts.

But we do have bootstrap styling classes for different alert types.

hudson-newey replied:
This looks quite good. I like your suggestion

![image](https://github.com/user-attachments/assets/a088f49d-a9a0-49dc-9cc6-918666ae3129)

atruskie commented on src/app/components/import-annotations/widgets/annotation-import-issue.component.html:
```diff
@@ -0,0 +1,104 @@
+    <div class="error-cards">
+      @if (hasEventErrors()) {
+        <div class="widget-card">
+          <h6 class="callout callout-danger">Event errors</h6>
```
```suggestion
          <h6 class="callout callout-danger">Individual errors</h6>
```

hudson-newey replied:
Nit: Why "Individual" instead of "Event".

I remember thinking about this terminology and used "Event" because it indicates that this error helper applies to the errors found in the "Identified Events" table.

atruskie replied:
Yeah so I was actually anti "Event" because we use annotation consistently. Individual sounded clearer, but you could sway me on s/event/annotation/

hudson-newey replied:
That actually makes a lot of sense.

If I was to replace this with "Annotation Errors", I'm thinking that I should change the heading for the "Identified Events" table to "Identified Annotations"?

hudson-newey replied:
Relevant commit:

[`d652cee` (#2303)](https://github.com/QutEcoacoustics/workbench-client/pull/2303/commits/d652ceeed094d5df5ed924f9d74e436fe6182b97)

atruskie replied:
Good point about also being consistent with the table heading - though i was leaning to renaming it something including _preview_.

hudson-newey replied:
> though i was leaning to renaming it something including preview

How about using "Annotation Preview" for the table's heading

atruskie commented on src/app/components/import-annotations/widgets/annotation-import-issue.component.html:
```diff
@@ -0,0 +1,104 @@
+
+          <p>
+            Event errors are broken down per-row in the "Identified Events"
+            table.
```
```suggestion
            Each detected annotation is shown in the preview table along with any detected errors
```

atruskie commented on src/app/components/import-annotations/widgets/annotation-import-issue.component.html:
```diff
@@ -0,0 +1,104 @@
+          <p>
+            Errors can arise due to invalid values, incorrect data types, or
+            referencing audio recordings that you don't have permission to
+            access.
```
This bit is a bit redundant.

```suggestion
            You may have to change your files to correct
            the errors so we can understand your data.
```

hudson-newey replied:
I like the idea of reducing the amount of text shown.

Although I'm a bit concerned about being vague.

---

In my mind, these error widgets are attempting to assist users in fixing their data.

e.g. Note that if you don't have access to an audio recording, the error that you will receive will be "audio recording not found", which doesn't provide much help with fixing the error.

---

After writing this out, I think the root cause is that some of the errors returned by the api are a bit vague.

E.g. we should return different error messages for unauthorized access and incorrect data types (e.g. strings instead of `int`'s)

atruskie replied:
>e.g. Note that if you don't have access to an audio recording, the error that you will receive will be "audio recording not found", which doesn't provide much help with fixing the error.

It _should_ actually tell you don't have permission - i specifically made the error distinct.

> Although I'm a bit concerned about being vague.

Fair, but then we should provide better error messages, customised to the errors we're seeing.

A basket list of possible errors is just text people don't want to read

hudson-newey replied:
![image](https://github.com/user-attachments/assets/28c84693-cf5e-4f7f-b73f-05021cc466ac)

You're correct, the permission failure is distinct per event returned.

However, the errors shown in the sidebar are using the errors returned from the response's `meta.error.info` object.

![image](https://github.com/user-attachments/assets/7bbfbaf8-b354-4b5e-a7d7-f78c05428ab3)

In this case, the error message for the file is "Validation failed".

atruskie replied:
Yeah and?

You have a function for searching through to get error types

atruskie commented on src/app/components/import-annotations/widgets/annotation-import-issue.component.html:
```diff
@@ -0,0 +1,104 @@
+            This includes other accounts uploading the same dataset. Please
+            ensure that the conflicting dataset has not previously been uploaded
+            by another person.
+          </p>
```
add contact us cta for help

hudson-newey replied:
![image](https://github.com/user-attachments/assets/17384ba4-774d-4e16-bf48-7459c6dde683)

atruskie commented on src/app/components/import-annotations/widgets/annotation-import-issue.component.html:
```diff
@@ -0,0 +1,104 @@
+
+          <p>
+            We encountered an unrecoverable error while processing your file. We
+            have been notified and will investigate this failure.
```
```suggestion
            We encountered an error while processing your file. We
            have been notified and will investigate this failure.
```

atruskie commented on src/app/components/import-annotations/widgets/annotation-import-issue.component.html:
```diff
@@ -0,0 +1,104 @@
+
+          <p>
+            This error can sometime arise if the structure of your dataset is
+            invalid or corrupted.
```
```suggestion
            This error can occur if the structure of your dataset is
            invalid or corrupted.
```

atruskie commented on src/app/components/import-annotations/widgets/annotation-import-issue.component.html:
```diff
@@ -0,0 +1,104 @@
+            If you have any further information, you can
+            <a [strongRoute]="contactUs.route">{{ contactUs.label }}</a> to help
+            resolve this bug faster.
+          </p>
```
we have a contact us cta like three times here, common template?

atruskie commented on src/app/directives/datatable/datatableConfig.ts:
```diff
@@ -0,0 +1,25 @@
+  public abstract rowHeight: InputSignal<DatatableComponent["rowHeight"]>;
+
+  public ngOnInit(): void {
+    this.datatableRef.nativeElement.classList.add("bootstrap");
```
is this common to both implementations of the config?

hudson-newey replied:
Yes, this should be consistent across all of our datatable configs because it applies the [bootstrap theming](https://swimlane.github.io/ngx-datatable/bootstrap) to the ngx-datatable.

Now that you bring it up, this should actually be refactored out of this directive and into the `_ngx-datatable.scss` styling.

hudson-newey replied:
![image](https://github.com/user-attachments/assets/0899de84-de40-4027-bd93-55a8bd3914bc)

atruskie commented on src/app/directives/datatable/datatableConfig.ts:
```diff
@@ -0,0 +1,25 @@
+    this.datatable.headerHeight = this.headerHeight();
+    this.datatable.summaryHeight = this.summaryHeight();
+
+    this.datatable.rowHeight = this.rowHeight();
```
bit useless  having a signal if we can only ever update the value once. Don't we want to listen to changes to these signals?

hudson-newey replied:
I have created a small test to see if the signals can be updated after creation.

https://github.com/user-attachments/assets/30252dc2-bfc0-4443-9236-12ac52ae39c8

![image](https://github.com/user-attachments/assets/1dcca3e8-cf84-41da-b32f-b9cc63319adf)
![image](https://github.com/user-attachments/assets/d5e2a8e1-3c7c-4c33-af76-3719c625f504)
![image](https://github.com/user-attachments/assets/0003a732-e1cb-4289-b14b-e44ad2da9ede)

---

I suspect one of two things is happening here:

- What I think is happening: Evaluating the signals value attaches it to the components `render3` change detection.
- or/and updating the `[footerHeight]` input is triggering the underlying `ngx-datatable` `input`.

My understanding of signals is that when they are invoked inside a `REACTIVE_NODE`, the reactive node's CD is automatically subscribed to the signal.

e.g.

```ts
class Component {
	public async update() {
		// do update
	}
}

const signalGetter = function(subscribers, value) {
	// "this" here refers to the "Component" or the nearest prototype
	subscribers.add(this.update);
	return value;
}

function signal<T>(initialValue: T) {
	const value: T = initialValue;
	const subscribers = new Set<(newValue: T) => void>([]);

	const notifySubscribers = (newValue) => {
		for (const subscriber of subscribers) {
			subscriber(newValue);
		}
	};

	const signalObj = function() {
		return signalGetter(subscribers, value);
	}
	signalObj.set = () => {
		notifySubscribers();
	};
	signalObj.subscribe = (callback: (value: T) => void)) {
		this.subscribers.add(callback);
	};

	return signalObj;
}
```

---

I'll obviously do some further investigation, but I think my mental model checks out.

atruskie replied:
I don't have the mental bandwidth for this right now, but i'm keen for you to walk me through it later

hudson-newey replied:
Discussed offline.

In my demonstration above, I demonstrated that updating the attributes does update the datatable, so there's no need for a subscription.

However, this is not because subscriptions automatically subscribed to components upon invokation.

The demo worked because it was triggering the old `@Input` change detection from the ngx-datatable component.

https://github.com/swimlane/ngx-datatable/blob/fe69d1e96b93bfc93e8e8ca9a1dc179db6593d05/projects/swimlane/ngx-datatable/src/lib/components/datatable.component.ts#L450

hudson-newey committed:
Add tests for compact datatable config

atruskie reviewed (changes requested):
add import annotations menu item to project menu page as well please (not withstanding scope) so users can find the item easily

hudson-newey committed:
Update src/app/components/import-annotations/widgets/annotation-import-issue.component.html

Co-authored-by: Anthony Truskinger <anthony.truskinger@gmail.com>

hudson-newey committed:
Update src/app/components/import-annotations/widgets/annotation-import-issue.component.html

Co-authored-by: Anthony Truskinger <anthony.truskinger@gmail.com>

hudson-newey committed:
Update src/app/components/import-annotations/widgets/annotation-import-issue.component.html

Co-authored-by: Anthony Truskinger <anthony.truskinger@gmail.com>

hudson-newey committed:
Update src/app/components/import-annotations/widgets/annotation-import-issue.component.html

Co-authored-by: Anthony Truskinger <anthony.truskinger@gmail.com>

hudson-newey committed:
Update src/app/components/import-annotations/widgets/annotation-import-issue.component.html

Co-authored-by: Anthony Truskinger <anthony.truskinger@gmail.com>

hudson-newey committed:
Update src/app/components/import-annotations/widgets/annotation-import-issue.component.html

Co-authored-by: Anthony Truskinger <anthony.truskinger@gmail.com>

hudson-newey committed:
Review feedback

hudson-newey committed:
Change "event" terminoligy to "annotation"

hudson-newey committed:
Revert information mock-up for non-saved warning

hudson-newey committed:
Move menu item to project list page

hudson-newey committed:
Style ngx-datatable through scss instead of CSSOM

hudson-newey committed:
Fix failing tests

hudson-newey committed:
Review feedback

hudson-newey committed:
Improve default.directive

hudson-newey committed:
Move annotation import route to project

This also makes the baw-error-card's error message scale to be 100% the
width/height of its host size.

This allows the error card to fill up the entire grid cell in the
annotation preview table.

hudson-newey committed:
Revert event error card changes

hudson-newey committed:
Errors fill entire annotation preview table cell

hudson-newey committed:
Update tests with annotation import project route

atruskie reviewed (approved):
💯
